### PR TITLE
[secure] add a vault network_timeout_ms

### PR DESCRIPTION
**Pull request by agclement**
## Motivation

add a configurable timeout to Vault.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

yes

## Test Plan

Don't break unit tests

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/libra/website, and link to your PR here.)

### CODE REVIEW

File: config/management/src/secure_backend.rs [hunk 1]
```diff
@@ -132,6 +132,8 @@ impl TryInto<config::SecureBackend> for SecureBackend {
                     server,
                     ca_certificate: certificate,
                     token: Token::FromDisk(PathBuf::from(token)),
+                    // TODO(davidiw):  parse the network_timeout_ms from the parameter list.
+                    network_timeout_ms: 30_000,
```
Comment (agclement) [inline config/management/src/secure_backend.rs:136]: @davidiw It's not clear to me how this field should go "up" from here.  A follow on commit/PR is needed to get this to the point of editable by an operator.

Comment (davidiw) [reply config/management/src/secure_backend.rs:136]: I don't think we need a todo here, I cannot imagine tweaks, but we always support adding optional parameters if we really want that.

File: config/src/config/secure_backend_config.rs [hunk 2]
```diff
@@ -45,6 +45,8 @@ pub struct VaultConfig {
     pub server: String,
     /// The authorization token for accessing secrets
     pub token: Token,
+    /// Network Timeout in milliseconds.
+    pub network_timeout_ms: u64,
```
Comment (davidiw) [inline config/src/config/secure_backend_config.rs:49]: nit alpha ordeing

**Comment by agclement** [comment]
@davidiw alpha nit fixed.   TODO eliminated;  it accomplished the primary purpose of determining further changes not needed :)

File: config/config-builder/src/key_manager_config.rs [hunk 3]
```diff
@@ -51,6 +53,7 @@ impl KeyManagerConfig {
             namespace: self.vault_namespace.clone(),
             server: self.vault_host.clone(),
             token: Token::FromConfig(self.vault_token.clone()),
+            network_timeout_ms: self.vault_network_timeout_ms,
```
Comment (davidiw) [inline config/config-builder/src/key_manager_config.rs:56]: maybe there's something quirky happening here but this isn't in alpha ordering

File: config/config-builder/src/validator_config.rs [hunk 4]
```diff
@@ -41,6 +41,7 @@ pub struct ValidatorConfig {
     pub safety_rules_token: Option<String>,
     pub seed: [u8; 32],
     pub template: NodeConfig,
+    pub vault_network_timeout_ms: u64,
```
Comment (davidiw) [inline config/config-builder/src/validator_config.rs:44]: I think we can just use the default, no need to expose this... how did we do the default for safety rules?

File: config/src/config/secure_backend_config.rs [hunk 5]
```diff
@@ -211,6 +216,7 @@ vault:
                 server: "127.0.0.1:8200".to_string(),
                 ca_certificate: None,
                 token: Token::FromDisk(PathBuf::from("/token")),
+                network_timeout_ms: 5_000,
```
Comment (davidiw) [inline config/src/config/secure_backend_config.rs:219]: same here

File: config/src/config/secure_backend_config.rs [hunk 6]
```diff
@@ -195,6 +199,7 @@ vault:
     server: "127.0.0.1:8200"
     token:
         from_config: "test"
+    network_timeout_ms: 5000
```
Comment (davidiw) [inline config/src/config/secure_backend_config.rs:202]: here

File: config/src/config/secure_backend_config.rs [hunk 7]
```diff
@@ -219,6 +225,7 @@ vault:
     server: "127.0.0.1:8200"
     token:
         from_disk: "/token"
+    network_timeout_ms: 5000
```
Comment (davidiw) [inline config/src/config/secure_backend_config.rs:228]: here

File: config/src/config/secure_backend_config.rs (hunk 5 above)
Comment (agclement) [reply config/src/config/secure_backend_config.rs:219]: The default is 30_000;  the 5_000 that shows up here is suitable for test usage only.

File: config/src/config/secure_backend_config.rs (hunk 6 above)
Comment (agclement) [reply config/src/config/secure_backend_config.rs:202]: Relevant if we're parsing the config from text;  either the field is present in vault_config in which case we should expose the value or not.

File: config/config-builder/src/validator_config.rs (hunk 4 above)
Comment (agclement) [reply config/config-builder/src/validator_config.rs:44]: We should use the default.  The question is if we want to hardcode the default value on line 226 or hardcode the default value on line 65.  I prefer hardcoding that value in `default()`, but leave that as a judgement call for you.

**Comment by agclement** [comment]
Applied one nit;  mild pushback on the other two.  One more round of discussion and I think we'll be done ...

**Comment by davidiw** [comment]
Ugh, looks like I dropped the ball here... will possibly commandeer unless you want to bump this again?

**Comment by agclement** [comment]
@davidiw  hopefully this resolves conflicts and passes tests.

File: config/config-builder/src/key_manager_config.rs [hunk 8]
```diff
@@ -15,6 +15,7 @@ pub struct KeyManagerConfig {
 
     pub vault_host: String,
     pub vault_namespace: Option<String>,
+    pub vault_network_timeout_ms: u64,
```
Comment (davidiw) [inline config/config-builder/src/key_manager_config.rs:18]: seems that this either needs to be optional or defined everywhere and one of the config doesn't have it defined... probably best to make this optional and then an unwrap_or(DEFAULT_VAULT_NETWORK_TIMEOUT_MS)

**Comment by agclement** [comment]
@davidiw problems solved!

File: config/src/config/secure_backend_config.rs [hunk 9]
```diff
@@ -45,6 +45,8 @@ pub struct VaultConfig {
     /// renewed up to that many seconds more. If this is not specified, the lease will not
     /// automatically be renewed.
     pub renew_ttl_secs: Option<u32>,
+    /// Network Timeout in milliseconds.
+    pub network_timeout_ms: u64,
```
Comment (davidiw) [inline config/src/config/secure_backend_config.rs:49]: I think making this non-optional is dangerous as it breaks existing configs that are outside the scope of this code base :/. I'll try to commandeer. Thanks for pushing on this!

**Comment by davidiw** [comment]
we have something in place now, closing out old PR

**Pull request closed without merging**